keps/sig-apps/0024-ttl-after-finish.md

```diff
@@ -0,0 +1,290 @@
+
+### Concrete Use Cases
+
+* [Kubeflow][] needs to clean up old finished Jobs (K8s Jobs, TF Jobs, Argo
+  workflows, etc.), see [#718][].
+
+* [Prow][] needs to clean up old completed Pods & finished Jobs. Currently implemented with Prow sinker.
+
+* [Apache Spark on Kubernetes][] needs proper cleanup of terminated Spark executor Pods.
+
+* Jenkins Kubernetes plugin creates slave pods that execute builds. It needs a better way to clean up old completed Pods.
+
+[Kubeflow]: https://github.com/kubeflow
+[#718]: https://github.com/kubeflow/tf-operator/issues/718
+[Prow]: https://github.com/kubernetes/test-infra/tree/master/prow
+[Apache Spark on Kubernetes]: http://spark.apache.org/docs/latest/running-on-kubernetes.html
+
+### Detailed Design 
+
+#### Feature Gate
+
+This will be launched as an alpha feature first, with feature gate
+`TTLAfterFinished`.
+
+#### API Object
+
+We will add the following API fields to `JobSpec` (`Job`'s `.spec`).
+
+```go
+type JobSpec struct {
+ 	// ttlSecondsAfterFinished limits the lifetime of a Job that has finished
+	// execution (either Complete or Failed). If this field is set, once the Job
+	// finishes, it will be deleted after ttlSecondsAfterFinished expires. When
+	// the Job is being deleted, its lifecycle guarantees (e.g. finalizers) will
+	// be honored. If this field is unset, ttlSecondsAfterFinished will not
+	// expire. If this field is set to zero, ttlSecondsAfterFinished expires
+	// immediately after the Job finishes.
+	// This field is alpha-level and is only honored by servers that enable the
+	// TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
+}
+```
+
+This allows Jobs to be cleaned up after they finish and provides time for
+asynchronous clients to observe Jobs' final states before they are deleted.
+
+
+Similarly, we will add the following API fields to `PodSpec` (`Pod`'s `.spec`).
+
+```go
+type PodSpec struct {
+ 	// ttlSecondsAfterFinished limits the lifetime of a Pod that has finished
+	// execution (either Succeeded or Failed). If this field is set, once the Pod
+	// finishes, it will be deleted after ttlSecondsAfterFinished expires. When
+	// the Pod is being deleted, its lifecycle guarantees (e.g. finalizers) will
+	// be honored. If this field is unset, ttlSecondsAfterFinished will not
+	// expire. If this field is set to zero, ttlSecondsAfterFinished expires
+	// immediately after the Pod finishes.
+	// This field is alpha-level and is only honored by servers that enable the
+	// TTLAfterFinished feature.
+	// +optional
+	TTLSecondsAfterFinished *int32
+}
+```
+
+##### Validation
+
+Because Job controller depends on Pods to exist to work correctly. In Job
+validation, `ttlSecondsAfterFinished` of its pod template shouldn't be set, to
+prevent users from breaking their Jobs. Users should set TTL seconds on a Job,
+instead of Pods owned by a Job.
```

Comment:
will job validation prevent this? how does this work with podgc today? does that mess up jobs?

Reply:
it's common for higher level resources to call generic pod spec validation, then add their own restrictions on top:

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/validation/validation.go#L121-L127

Reply:
Thanks for the advice. Added validation for restart policy.